messenger/src/pulsar_messenger.rs

```diff
@@ -0,0 +1,217 @@
+#![cfg(feature = "pulsar")]
+
+use {
+    crate::{error::MessengerError, Messenger, MessengerConfig},
+    async_mutex::Mutex,
+    async_trait::async_trait,
+    futures::TryStreamExt,
+    log::*,
+    pulsar::{consumer::Message, Authentication, Consumer, Producer, Pulsar, TokioExecutor},
+    std::sync::Arc,
+    std::{
+        collections::HashMap,
+        fmt::{Debug, Formatter},
+    },
+};
+
+#[derive(Default)]
+pub struct PulsarMessenger {
+    connection: Option<Pulsar<TokioExecutor>>,
+    producers: HashMap<&'static str, Producer<TokioExecutor>>,
+    pub consumers: HashMap<&'static str, Arc<Mutex<Consumer<Vec<u8>, TokioExecutor>>>>,
+    max_buffer_size: HashMap<&'static str, usize>,
+    pub acquired_messages: HashMap<&'static str, Vec<Message<Vec<u8>>>>,
+}
+
+const PULSAR_CON_STR: &str = "pulsar_connection_str";
+const PULSAR_AUTH_TOKEN: &str = "pulsar_auth_token";
+
+#[async_trait]
+impl Messenger for PulsarMessenger {
+    /// Create new Pulsar connection for future topics
+    async fn new(config: MessengerConfig) -> Result<Self, MessengerError> {
+        let uri = config
+            .get(&*PULSAR_CON_STR)
+            .and_then(|u| u.clone().into_string())
+            .ok_or(MessengerError::ConfigurationError {
+                msg: format!("Connection String Missing: {}", PULSAR_CON_STR),
+            })?;
+
+        let mut builder = Pulsar::builder(uri, TokioExecutor);
+
+        if let Some(token) = config.get(&*PULSAR_AUTH_TOKEN) {
+            let authentication = Authentication {
+                name: String::from("token"),
+                data: token.clone().into_string().unwrap().into_bytes(),
+            };
+            builder = builder.with_auth(authentication);
+        }
+
+        let pulsar: Pulsar<_> = builder.build().await.unwrap();
+
+        Ok(Self {
+            connection: Some(pulsar),
+            producers: HashMap::<&'static str, Producer<TokioExecutor>>::default(),
+            consumers:
+                HashMap::<&'static str, Arc<Mutex<Consumer<Vec<u8>, TokioExecutor>>>>::default(),
+            max_buffer_size: HashMap::<&'static str, usize>::default(),
+            acquired_messages: HashMap::<&'static str, Vec<Message<Vec<u8>>>>::default(),
+        })
+    }
+
+    /// Create new Producer for Pulsar topic
+    async fn add_stream(&mut self, stream_key: &'static str) {
+        if self.producers.contains_key(stream_key) {
+            error!("Stream {stream_key} already exists");
+            return;
+        }
+
+        let producer = self
+            .connection
+            .as_mut()
+            .unwrap()
+            .producer()
+            .with_topic(stream_key)
+            .build()
+            .await
+            .unwrap();
```

Comment:
If building the `producer` or `consumer` can have more interesting failures at runtime that might be handled differently, we can consider changing the interface to `add_stream()` to return a `MessengerError`.  I think with Redis Streams the only error we considered was "stream already existed" which is not really an actionable error.

But if you think there's a good reason to notify the caller from these producer/consumer build functions, we can make the change.  If it is not likely to fail in normal runtime, feel free to also leave it as `unwrap()`.

Reply:
Sure it makes sense, thank you to notice it
Creating Consumer or Producer in Pulsar can return different errors with different reasons
Will add it